README.md
# bombPartY
A python/pygame port of the "BombParty" word game, to be run as a standalone.

## Usage
Server and players must have version 2+ of the pygame package installed.

The server for the game is run by
	python3 game_server.py (ip) (port) (mode) (savefile)
ip and port specify where the server will listen for connections from game clients. mode should always be 'new' for now. savefile specifies where the game logs will be stored.

Once the server is running, the players connect by
	python3 game_client.py (ip) (port) (username)
ip and port are the same as that specified by the server, and username is a distinguishing display name, which will automatically be limited to 20 characters.


CHANGELOG
BombPartY Changelog
v0.3.1
- Fixed bug where clearing partial entries on player death caused it to be constantly cleared

v0.3
- Can now play multiple games without restarting the program
- Fixed bug where partial entries carried over on player death

v0.2.2
- Entered players automatically switch focus to game box when the game starts

v0.2.1
- Fixed screen only updating when the window receives an event
- Refactored networking code and some game code for clarity

v0.2
- Updated networking handling in game_client.py and game_server.py to allow for latencies greater than localhost


chat.py
#BombPartY v0.3.1 - a PyGame port of the classic wordgame
#Copyright (C) 2023 Daniel Bassett

#This program is free software: you can redistribute it and/or modify
#it under the terms of the GNU General Public License as published by
#the Free Software Foundation, either version 3 of the License, or
#(at your option) any later version.

#This program is distributed in the hope that it will be useful,
#but WITHOUT ANY WARRANTY; without even the implied warranty of
#MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#GNU General Public License for more details.

#You should have received a copy of the GNU General Public License
#along with this program.  If not, see <https://www.gnu.org/licenses/>.

import pygame

class message:
	def __init__(self, message, username):
		self.message = message
		self.username = username
	
	def toDisplay(self):
		return self.username + ": " + self.message
		
	def drawText(self, font, surface, y):
		text = self.toDisplay()
		while len(text) > 0:
			i = 0
			while i < len(text) and font.size(text[:i])[0] < 350: #determine the maximum text that will fit on the line
				i += 1
			
			if font.size(text[:i])[0] >= 350:
				i -= 1
				i = text.rfind(" ", 0, i) + 1
			
			text_surface = font.render(text[:i], True, (255,255,255))
			surface.blit(text_surface, (5, y))
			
			y += font.size(text[:i])[1] + 5
			text = text[i:]
		
		return y
			

class chatlog:
	def __init__(self):
		self.messages = []
		self.bottom = True
		self.scrollY = 0
		self.height = 540
	
	def drawHeight(self, font):
		height = 0
		dummy_surface = pygame.Surface((360, 360))
		for msg in self.messages:
			height += msg.drawText(font, dummy_surface, 0) + 5
		return max(height, 540)
	
	def packAll(self):
		chat_dict = {'num_messages': len(self.messages),
					'messages': [self.messages[i].message for i in range(len(self.messages))],
					'usernames': [self.messages[i].username for i in range(len(self.messages))],
					'data': "chat"}
		
		return dict(type="text/json", encoding="utf-8", content = chat_dict)
	
	def depackAll(self, decode, font):
		if decode.get('num_messages') > len(self.messages):
			for i in range(len(self.messages), decode.get('num_messages')):
				self.messages.append(message(decode.get('messages')[i], decode.get('usernames')[i]))
		
		if self.bottom:
			self.height = self.drawHeight(font)
			self.scrollY = self.height - 540
		else:
			self.height = self.drawHeight(font)


game.py
#BombPartY v0.3.1 - a PyGame port of the classic wordgame
#Copyright (C) 2023 Daniel Bassett

#This program is free software: you can redistribute it and/or modify
#it under the terms of the GNU General Public License as published by
#the Free Software Foundation, either version 3 of the License, or
#(at your option) any later version.

#This program is distributed in the hope that it will be useful,
#but WITHOUT ANY WARRANTY; without even the implied warranty of
#MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#GNU General Public License for more details.

#You should have received a copy of the GNU General Public License
#along with this program.  If not, see <https://www.gnu.org/licenses/>.

import random
import json
import time

import user

class game:
	def __init__(self):
		self.started = False
		self.ended = False
		self.end_time = 0
		
		self.players = []
		self.alive_players = 0
		
		self.prompt = ""
		self.prompt_start_player = 0
		self.prompt_start_time = 0
		self.current_player = 0
		
		self.current_entry = ""
		
		self.used_words = []
		self.dictionary = []
		
		dictionary_stream = open("dictionary/english_small.txt","r")
		for word in dictionary_stream:
			self.dictionary.append(word[:-1]) #Removes EOL character
		dictionary_stream.close()
		
		self.prompt_difficulty = 0
	
	#Generate prompt
	def generatePrompt(self):
		if self.prompt_difficulty == 0: #Standard
			wordNumber = len(self.dictionary)
			while wordNumber == len(self.dictionary) or self.dictionary[wordNumber] in self.used_words:
				wordNumber = random.randrange(0, len(self.dictionary))
	
			wordLength = len(self.dictionary[wordNumber])
			promptLength = random.randint(2,3)
			promptStart = random.randrange(0, wordLength-promptLength)
			
			self.prompt = self.dictionary[wordNumber][promptStart:promptStart+promptLength]
			
		self.prompt_start_time = time.time_ns()
		self.prompt_start_player = self.current_player
		
	#Check validity of user entry	
	def checkForPrompt(self, word):
		found = False
		for i in range(len(word) - len(self.prompt)+1):
			if word[i:i+len(self.prompt)] == self.prompt:
				found = True
		return found
	
	def checkDictionary(self, word):
		lower = 0
		upper = len(self.dictionary)
		
		#Checks if word is in dictionary by binary search
		while upper - lower > 1:
			if word < self.dictionary[(upper+lower)>>1]:
				upper = (upper+lower)>>1
			else:
				lower = (upper+lower)>>1
		
		return word == self.dictionary[lower]
	
	def findNextAlive(self):
		#Sets the current player to the next alive player
		#Returns True if a new prompt is needed.
		wrapped = (len(self.players) == 1)
		self.current_player = (self.current_player + 1) % len(self.players) #move at least one player
		if self.current_player == self.prompt_start_player:
			wrapped = True
		
		while self.players[self.current_player].lives == 0: #skip to the next if that player is dead
			self.current_player = (self.current_player + 1) % len(self.players)
			if self.current_player == self.prompt_start_player:
				wrapped = True
		
		return wrapped
	
	#Network helper functions
	def packAll(self, seatNum, letters): #packs into json for transfer to player in seat seatNum
		game_dict = {'started': self.started, 
					'ended': self.ended,
					'num_players': len(self.players), 
					'player_lives': [self.players[i].lives for i in range(len(self.players))],
					'player_usernames': [self.players[i].username for i in range(len(self.players))],
					'current_player': self.current_player,
					'prompt': self.prompt,
					'used_words': self.used_words,
					'timestamp': time.time_ns(),
					'data': "full",
					'your_seat': seatNum,
					'your_letters': letters,
					'current_entry': self.current_entry}
		
		return dict(type="text/json", encoding="utf-8", content = game_dict)
	
	def depackAll(self, decode, player):
		if self.started != decode.get('started') and player.playing: #When game starts, set priority to game box
			player.chatting = False
			
		self.started = decode.get('started')
		self.ended = decode.get('ended')
		self.current_player = decode.get('current_player')
		self.prompt = decode.get('prompt')
		self.used_words = decode.get('used_words')
		self.current_entry = decode.get('current_entry')
		player.seat = decode.get('your_seat')
		player.letters = decode.get('your_letters')
		if decode.get('num_players') > len(self.players): #add new players
			for i in range(len(self.players), decode.get('num_players')):
				self.players.append(user.user(-1))
				self.players[-1].playing = True
				self.players[-1].username = decode.get('player_usernames')[i]
				
		elif decode.get('num_players') < len(self.players): #game restarted
			self.players = self.players[:decode.get('num_players')]
			player.playing = False
		
		
		for i in range(len(self.players)): #update lives
			self.players[i].lives = decode.get('player_lives')[i]
		
		#reset word on death
		if player.seat > -1 and decode.get('player_lives')[player.seat] < player.lives:
			player.word = ""
			player.lives = decode.get('player_lives')[player.seat]


user.py
#BombPartY v0.3.1 - a PyGame port of the classic wordgame
#Copyright (C) 2023 Daniel Bassett

#This program is free software: you can redistribute it and/or modify
#it under the terms of the GNU General Public License as published by
#the Free Software Foundation, either version 3 of the License, or
#(at your option) any later version.

#This program is distributed in the hope that it will be useful,
#but WITHOUT ANY WARRANTY; without even the implied warranty of
#MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#GNU General Public License for more details.

#You should have received a copy of the GNU General Public License
#along with this program.  If not, see <https://www.gnu.org/licenses/>.

import chat

class user:
	def __init__(self, connection_no):
		self.connection_no = connection_no
		self.username = ""
		self.playing = False
		
		self.chat_message = chat.message("", "")
		self.chatting = True

		self.lives = 2
		self.letters = {"A": True, "B": True, "C": True, "D": True, "E": True, "F": True, "G": True, "H": True, "I": True, "J": True, "K": True, "L": True, "M": True, "N": True, "O": True, "P": True, "Q": True, "R": True, "S": True, "T": True, "U": True, "V": True, "W": True, "Y": True} #True if still to be used
		
		self.seat = -1
		
		self.word = ""
		
		self.prompt_received = 0 #stores time that the prompt was received
	
	def updateLetters(self, word):
		for letter in word:
			if letter == "X" or letter == "Z":
				continue
			self.letters[letter] = False
		
		flip = True
		for letter in self.letters:
			if self.letters[letter]:
				flip = False
				break
		
		if flip:
			for letter in self.letters:
				self.letters[letter] = True
			
			if self.lives < 3:
				self.lives += 1


networking.py
#BombPartY v0.3.1 - a PyGame port of the classic wordgame
#Copyright (C) 2023 Daniel Bassett

#This program is free software: you can redistribute it and/or modify
#it under the terms of the GNU General Public License as published by
#the Free Software Foundation, either version 3 of the License, or
#(at your option) any later version.

#This program is distributed in the hope that it will be useful,
#but WITHOUT ANY WARRANTY; without even the implied warranty of
#MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#GNU General Public License for more details.

#You should have received a copy of the GNU General Public License
#along with this program.  If not, see <https://www.gnu.org/licenses/>.

"""Module containing the structure used for a socket. Transfers json files."""

import socket
import selectors
import sys
import struct
import json

def start_connection(host, port, selector):
    """Connects the client to the server."""
    addr = (host, port)
    print("starting connection to", addr)
    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    sock.setblocking(False)
    sock.connect_ex(addr)
    events = selectors.EVENT_WRITE | selectors.EVENT_READ
    message = Message(selector, sock, addr)
    selector.register(sock, events, data=message)
    #print("Connected successfully.")
    return message

def open_connection(host, port, selector):
    """Opens the server to client connections."""
    lsock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    # Avoid bind() exception: OSError: [Errno 48] Address already in use
    lsock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    lsock.bind((host, port))
    lsock.listen()
    print("listening on", (host, port))
    lsock.setblocking(False)
    selector.register(lsock, selectors.EVENT_READ, data=None)

def accept_wrapper(sock, selector):
    """Accept connections from client, add to selector."""
    conn, addr = sock.accept()  # Should be ready to read
    print("accepted connection from", addr)
    conn.setblocking(False)
    message = Message(selector, conn, addr)
    selector.register(conn, selectors.EVENT_READ | selectors.EVENT_WRITE, data=message)
    return message

class Message:
    """Class that handles the sending and receiving of json messages. Data of a socket."""
    def __init__(self, selector, sock, addr):
        self.selector = selector
        self.sock = sock
        self.addr = addr
        self._recv_buffer = b""
        self._send_buffer = b""

        self._jsonheader_len = None
        self.jsonheader = None
        self.received = None

        self.to_send = {}

    def _read(self):
        try:
            #print("attempting to read data")
            data = self.sock.recv(4096)
            #print("successfully read data")
        except BlockingIOError:
            print("might be blocking issue (read)")
            pass
        else:
            if data:
                self._recv_buffer += data
                print(data)
            else:
                raise RuntimeError("Peer closed.")

    def _write(self):
        if self._send_buffer:
            print("sending", repr(self._send_buffer))
            try:
                sent = self.sock.send(self._send_buffer)
            except BlockingIOError:
                print("might be blocking issue (write)")
                pass
            else:
                self._send_buffer = self._send_buffer[sent:]

    def _json_encode(self, obj, encoding):
        return json.dumps(obj, ensure_ascii=False).encode(encoding)

    def _json_decode(self, json_bytes, encoding):
        return json.loads(json_bytes.decode(encoding))

    def _create_message(self, *, content_bytes, content_type, content_encoding):
        jsonheader = {
            "byteorder": sys.byteorder,
            "content-type": content_type,
            "content-encoding": content_encoding,
            "content-length": len(content_bytes),
        }
        jsonheader_bytes = self._json_encode(jsonheader, "utf-8")
        message_hdr = struct.pack(">H", len(jsonheader_bytes))
        message = message_hdr + jsonheader_bytes + content_bytes
        return message

    def process_events(self, mask):
        """Check if the socket should be read or written to."""
        if mask & selectors.EVENT_READ:
            move = self.read()
            return move
        if mask & selectors.EVENT_WRITE:
            self.write()

    def read(self):
        """Reading process."""
        self._read()

        if self._jsonheader_len is None:
            self.process_protoheader()

        if self._jsonheader_len is not None:
            if self.jsonheader is None:
                self.process_jsonheader()

        if self.jsonheader:
            if self.received is None:
                move = self.process_content()
                return move
        return None

    def write(self):
        """Writing process."""
        if self.to_send:
            self.queue_message()

        self._write()

    def queue_message(self):
        """Package messages to be sent and move them to the buffer."""
        content = self.to_send["content"]
        content_type = self.to_send["type"]
        content_encoding = self.to_send["encoding"]
        if content_type == "text/json":
            req = {
                "content_bytes": self._json_encode(content, content_encoding),
                "content_type": content_type,
                "content_encoding": content_encoding,
            }

        message = self._create_message(**req)
        self._send_buffer += message
        self.to_send = None

    def process_protoheader(self):
        """Process the protoheader, which contains the length of the jsonheader."""
        protoheader_len = 2

        if len(self._recv_buffer) >= protoheader_len:
            self._jsonheader_len = struct.unpack(
                ">H", self._recv_buffer[:protoheader_len]
            )[0]
            self._recv_buffer = self._recv_buffer[protoheader_len:]

    def process_jsonheader(self):
        """Process the jsonheader, which contains the length of the content."""
        jsonheader_len = self._jsonheader_len
        if len(self._recv_buffer) >= jsonheader_len:
            self.jsonheader = self._json_decode(
                self._recv_buffer[:jsonheader_len], "utf-8"
            )
            self._recv_buffer = self._recv_buffer[jsonheader_len:]
        for required_header in ("byteorder", "content-length", "content-type", "content-encoding"):
            if required_header not in self.jsonheader:
                raise ValueError(f'Missing required header "{required_header}".')

    def process_content(self):
        """Process the actual content of the message."""
        content_len = self.jsonheader["content-length"]
        if not len(self._recv_buffer) >= content_len:
            return
        data = self._recv_buffer[:content_len]
        self._recv_buffer = self._recv_buffer[content_len:]

        if self.jsonheader["content-type"] == "text/json":
            encoding = self.jsonheader["content-encoding"]
            self.received = self._json_decode(data, encoding)
            print("received response", repr(self.received))
            move = self.received
            self._jsonheader_len = None
            self.jsonheader = None
            self.received = None
            return move
        return None

    def close(self):
        """Close the connection."""
        print("closing connection to", self.addr)
        try:
            self.selector.unregister(self.sock)
        except Exception as err:
            print(
                "error: selector.unregister() exception for",
                f"{self.addr}: {repr(err)}",
            )

        try:
            self.sock.close()
        except OSError as err:
            print(
                "error: socket.close() exception for",
                f"{self.addr}: {repr(err)}",
            )
        finally:
            self.sock = None

    def nothing_to_send(self):
        """Checks whether the message is sending anything."""
        if self.to_send:
            return False
        elif self._send_buffer:
            return False
        return True
        
    def readAll(self):
        decodes = []
        events = self.selector.select(timeout=1)
        for key, mask in events:
            message = key.data
            decode = False
            try: 
                potential_decode = message.process_events(mask)
                if potential_decode:
                    decode = potential_decode

            except Exception:
                message.close()
            
            if decode:
                decodes.append(decode)
                
        return decodes
    
    def sendAll(self):
        while not self.nothing_to_send:
            self.readAll()
    
    


game_client.py
#BombPartY v0.3.1 - a PyGame port of the classic wordgame
#Copyright (C) 2023 Daniel Bassett

#This program is free software: you can redistribute it and/or modify
#it under the terms of the GNU General Public License as published by
#the Free Software Foundation, either version 3 of the License, or
#(at your option) any later version.

#This program is distributed in the hope that it will be useful,
#but WITHOUT ANY WARRANTY; without even the implied warranty of
#MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#GNU General Public License for more details.

#You should have received a copy of the GNU General Public License
#along with this program.  If not, see <https://www.gnu.org/licenses/>.

import sys
import networking
import selectors
import pygame
import time

import user
import game
import chat

FPS = 60
letter_list = ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'Y']

pygame.init()
pygame.event.set_allowed([pygame.QUIT, pygame.KEYDOWN, pygame.MOUSEBUTTONDOWN, pygame.MOUSEWHEEL]) 

def main():
	"""Plays games of bombparty"""

	#License info
	print("bombPartY v0.3.1 Copyright (C) 2023 Daniel Bassett")
	print("This program comes with ABSOLUTELY NO WARRANTY.")
	print("This is free software, and you are welcome to redistribute it under certain conditions.")
	print("See the GNU General Public License for more details.")

	if len(sys.argv) != 4:
		print("Usage:", sys.argv[0], "<host> <port> <username>")
		sys.exit(1)
	host, port, username = sys.argv[1], int(sys.argv[2]), sys.argv[3][:20] #TODO: sanitise input
	username = sys.argv[3]
    
	#Initiate connection
	sel = selectors.DefaultSelector()
	msg = networking.start_connection(host, port, sel)
    
	#Send file validation (TODO)
	
	me = False
	
	current_game = game.game()
	game_chat = chat.chatlog()
	
	#REMOVE, DEBUGGING ONLY
	#me.playing = True
	#me = user.player(username)
	#current_game = game.game()
	
	#Initialise window
	screen = pygame.display.set_mode((1440, 720))
	pygame.display.set_caption("bombPartY") #Window title
	screen.fill((181,131,90))
	clock = pygame.time.Clock()
	
	
	join_rect = pygame.Rect(620, 600, 100, 32) #join game button
	
	prompt_rect = pygame.Rect(220, 500, 640, 32) #display current prompt
	
	answer_rect = pygame.Rect(220, 600, 640, 32) #Display currently typed answer
	
	player_rect = pygame.Rect(0, 0, 1000, 360) #Display player list
	
	letter_rect = pygame.Rect(1000,0,80,720) #Display player's letters
	
	chat_rect = pygame.Rect(1080, 0, 360, 540) #Display chat messages
	chat_message_rect = pygame.Rect(1080, 540, 360, 180) #Display currently typed message
	pygame.draw.rect(screen, pygame.Color(90,65,45), chat_rect)
	pygame.draw.rect(screen, pygame.Color(45,35,25), chat_message_rect)
	
	pygame.display.flip()
	
	#Set up text
	base_font = pygame.font.Font(None, 32)
	chat_font = pygame.font.Font(None, 32)
	
	#Main loop
	running = True
	screen_update = True
	connecting_drawn = False
	while running:
		#Networking updates
		decodes = msg.readAll()
		for decode in decodes:
			message_type = decode.get('data')
			print(message_type)
						
			if message_type == "connection": #received connection number
				connection_no = decode.get('connection_no')
				me = user.user(connection_no)
				me.chat_message.username = username
						
				msg.to_send = dict(type="text/json", encoding="utf-8", content={'username': username, 'timestamp': time.time_ns(), 'data': 'username', 'connection_no': connection_no})
				msg.write()
					
			elif message_type == "full":
				#print("this should happen") #it isn't happening for god knows what reason. it's showing up as sending.
				current_game.depackAll(decode, me)
				screen_update = True
						
			elif message_type == "chat":
				game_chat.depackAll(decode, chat_font)
				screen_update = True
			
			elif message_type == "game_start":
				if me.playing:
					me.chatting = False
		
		msg.sendAll()
	
		#Handle quit, typing etc.
		for event in pygame.event.get():
			if event.type == pygame.QUIT:
				running = False
			
			if me:
				if event.type == pygame.KEYDOWN:
					if me.chatting or not me.playing: #chat entry
						if event.key == pygame.K_BACKSPACE:
							me.chat_message.message = me.chat_message.message[:-1]
							screen_update = True
						
						elif event.key == pygame.K_RETURN or event.key == pygame.K_KP_ENTER: #send message
							msg.to_send = dict(type="text/json", encoding="utf-8", content={'connection_no': me.connection_no, 'message': me.chat_message.message, 'data': "chat"})
							msg.write()
							me.chat_message.message = ""
							screen_update = True
						
						elif len(me.chat_message.message) < 100: #TODO: sanitise input
							me.chat_message.message += event.unicode
							screen_update = True
					
					elif current_game.current_player == me.seat: #word entry
						if event.key == pygame.K_BACKSPACE:
							me.word = me.word[:-1]
							msg.to_send = dict(type="text/json", encoding="utf-8", content={'connection_no': me.connection_no, 'word': me.word, 'data': "type"})
							msg.write()
							
							screen_update = True
						
						elif event.key == pygame.K_RETURN or event.key == pygame.K_KP_ENTER:
							#TODO: add game stuff?
							msg.to_send = dict(type="text/json", encoding="utf-8", content={'connection_no': me.connection_no, 'word': me.word, 'data': "solve"})
							msg.write()
							me.word = ""
							screen_update = True
						
						elif len(me.word) < 30 and pygame.K_a <= event.key and event.key <= pygame.K_z: #TODO: better sanitising e.g. to stop ctrl+C stuff giving funny characters
							me.word += event.unicode
							me.word = me.word.upper()
							msg.to_send = dict(type="text/json", encoding="utf-8", content={'connection_no': me.connection_no, 'word': me.word, 'data': "type"})
							msg.write()
							
							screen_update = True
				
				elif event.type == pygame.MOUSEBUTTONDOWN:
					mouseX, mouseY = event.pos
					if current_game.started:
						if me.playing:
							if chat_message_rect.collidepoint(mouseX, mouseY):
								me.chatting = True
								
							elif answer_rect.collidepoint(mouseX, mouseY):
								me.chatting = False
							
					else:
						if not me.playing:
							if join_rect.collidepoint(mouseX, mouseY):
								me.playing = True
								msg.to_send = dict(type="text/json", encoding="utf-8", content={'connection_no': me.connection_no, 'data': 'game_join'})
								msg.write()
						elif me.seat == 0:
							if join_rect.collidepoint(mouseX, mouseY):
								msg.to_send = dict(type="text/json", encoding="utf-8", content={'connection_no': me.connection_no, 'data': 'game_start'})
								msg.write()
				
				elif event.type == pygame.MOUSEWHEEL:
					game_chat.scrollY -= 10*event.y #scroll by the given amount
					if event.y > 0: #No longer snap to the bottom
						game_chat.bottom = False
						
					if game_chat.scrollY < 0: #prevent scrolling off the top of chat
						game_chat.scrollY = 0 
					
					elif game_chat.scrollY > game_chat.height - 540:
						game_chat.scrollY = game_chat.height - 540
						game_chat.bottom = True #snap to the bottom
					
					screen_update = True
		
		if me:
			#Update screen
			if screen_update:
				screen_update = False
				
				if current_game.started:
					if not current_game.ended:
						pygame.draw.rect(screen, pygame.Color(181,131,90), prompt_rect)
						prompt_surface = base_font.render("Quick! Type an English word containing " + current_game.prompt, True, (255,255,255))
						screen.blit(prompt_surface, (prompt_rect.x, prompt_rect.y))
					
						if current_game.current_player == me.seat:
							#Draw answer entry rectangle
							pygame.draw.rect(screen, pygame.Color(45,35,25), answer_rect)
							answer_surface = base_font.render(me.word, True, (255,255,255))
							screen.blit(answer_surface, (answer_rect.x+20,answer_rect.y+5))
						else:
							pygame.draw.rect(screen, pygame.Color(181,131,90), answer_rect)
							answer_surface = base_font.render(me.word, True, (255,255,255))
							screen.blit(answer_surface, (answer_rect.x+20,answer_rect.y+5))
					
					else:
						pygame.draw.rect(screen, pygame.Color(181,131,90), prompt_rect)
						pygame.draw.rect(screen, pygame.Color(181,131,90), answer_rect)
						if current_game.current_player == me.seat:
							prompt_surface = base_font.render("Congratulations! You have won.", True, (255,255,255))
						else:
							prompt_surface = base_font.render("Unlucky! The winner is: " + str(current_game.players[current_game.current_player].username), True, (255,255,255))
						
						screen.blit(prompt_surface, (prompt_rect.x, prompt_rect.y))
							
				elif not me.playing:
					#Draw join button
					pygame.draw.rect(screen, pygame.Color(45,35,25), join_rect)
					join_surface = base_font.render("Join!", True, (255,255,255))
					screen.blit(join_surface, (join_rect.x+10, join_rect.y+5))
					
					pygame.draw.rect(screen, pygame.Color(181,131,90), prompt_rect)
				
				else:
					pygame.draw.rect(screen, pygame.Color(181,131,90), prompt_rect) #
				
					#Waiting to play
					if me.seat == 0:
						#print(len(current_game.players))
						pygame.draw.rect(screen, pygame.Color(45,35,25), join_rect)
						join_surface = base_font.render("Start!", True, (255,255,255))
						screen.blit(join_surface, (join_rect.x+10, join_rect.y+5))
					else:
						pygame.draw.rect(screen, pygame.Color(181, 131, 90), join_rect)
				
				#Draw previous chat messages
				chat_surface_full = pygame.Surface((360, game_chat.height))
				pygame.draw.rect(chat_surface_full, pygame.Color(90,65,45), (pygame.Rect(0,0, chat_rect.w, chat_rect.h)))
				y = 0
				
				for chatMsg in game_chat.messages: #Draws the full chat history
					y = chatMsg.drawText(chat_font, chat_surface_full, y) + 5
				
				chat_display_rect = pygame.Rect(0, game_chat.scrollY, 360,540) #Selects the part of the chat for the player's current scrolling value
				chat_surface_display = chat_surface_full.subsurface(chat_display_rect) 
				
				screen.blit(chat_surface_display, (chat_rect.x, chat_rect.y))
				
				#Draw current chat message
				chat_message_surface = pygame.Surface((360, 180))
				pygame.draw.rect(chat_message_surface, pygame.Color(45,35,25), pygame.Rect(0,0, chat_message_rect.w, chat_message_rect.h)) #why is this not doing anything
				me.chat_message.drawText(chat_font, chat_message_surface, 5)
				screen.blit(chat_message_surface, (chat_message_rect.x, chat_message_rect.y))
				
				#Draw players remaining letters
				pygame.draw.rect(screen, pygame.Color(181,131,90), letter_rect)
				for i in range(24):
					letter_surface = base_font.render(letter_list[i], True, (255,255,255))
					x = 1000
					if i >= 12:
						x += 40
						
					if me.letters[letter_list[i]]:
						screen.blit(letter_surface, (x, 40*(i%12)))
				
				#List entrant names
				pygame.draw.rect(screen, pygame.Color(181,131,90), player_rect)
				for i in range(len(current_game.players)):
					player_string = str(i)+" "+current_game.players[i].username+" "+str(current_game.players[i].lives)
					if i == current_game.current_player:
						player_string += " < " + current_game.current_entry
					entrant_surface = base_font.render(player_string, True, (255,255,255))
					screen.blit(entrant_surface, (0,32*i))
				
				pygame.display.flip()
		

		if not connecting_drawn:
			connecting_surface = base_font.render("Connecting...", True, (255,255,255))
			screen.blit(connecting_surface, (0,0))
			connecting_drawn = True
		
		#New frame			
		clock.tick(FPS)
	

if __name__ == "__main__":
	main()
